components/ai_chat/core/browser/model_service.cc

```diff
@@ -0,0 +1,356 @@
+      auto model = mojom::Model::New();
+      model->key = "chat-basic";
+      model->display_name = "Llama 3 8b";
+      model->options =
+          mojom::ModelOptions::NewLeoModelOptions(std::move(options));
+
+      models.push_back(std::move(model));
+    }
+
+    return models;
+  }());
+
+  return *kModels;
+}
+
+base::Value::Dict GetModelDict(mojom::ModelPtr model) {
+  base::Value::Dict model_dict = base::Value::Dict();
+  model_dict.Set(kCustomModelItemLabelKey, model->display_name);
+  model_dict.Set(
+      kCustomModelItemModelKey,
+      model->options->get_custom_model_options()->model_request_name);
+  model_dict.Set(kCustomModelItemEndpointUrlKey,
+                 model->options->get_custom_model_options()->endpoint.spec());
+  model_dict.Set(kCustomModelItemApiKey,
+                 model->options->get_custom_model_options()->api_key);
+  model_dict.Set(kCustomModelItemKey, model->key);
+  return model_dict;
+}
+}  // namespace
+
+ModelService::ModelService(PrefService* prefs_service)
+    : pref_service_(prefs_service) {
+  InitModels();
+
+  pref_change_registrar_.Init(pref_service_);
+}
+
+ModelService::~ModelService() = default;
+
+// static
+const mojom::Model* ModelService::GetModelForTesting(std::string_view key) {
+  const std::vector<mojom::ModelPtr>& all_models = GetLeoModels();
+
+  auto match_iter = std::find_if(
+      all_models.cbegin(), all_models.cend(),
+      [&key](const mojom::ModelPtr& model) { return model->key == key; });
+  if (match_iter != all_models.cend()) {
+    return &*match_iter->get();
+  }
+
+  return nullptr;
+}
+
+void ModelService::InitModels() {
+  // Get leo and custom models
+  const std::vector<mojom::ModelPtr>& leo_models = GetLeoModels();
+  std::vector<mojom::ModelPtr> custom_models = GetCustomModelsFromPrefs();
+
+  // Reserve space in the combined models vector
+  models_.clear();
+  models_.reserve(leo_models.size() + custom_models.size());
+
+  // Ensure we return only in intended display order
+  std::transform(leo_models.cbegin(), leo_models.cend(),
+                 std::back_inserter(models_),
+                 [](const mojom::ModelPtr& model) { return model.Clone(); });
+
+  std::transform(custom_models.cbegin(), custom_models.cend(),
+                 std::back_inserter(models_),
+                 [](const mojom::ModelPtr& model) { return model.Clone(); });
+
+  for (auto& obs : observers_) {
+    obs.OnModelListUpdated();
+  }
+}
+
+const std::vector<mojom::ModelPtr>& ModelService::GetModels() {
+  return models_;
+}
+
+const mojom::Model* ModelService::GetModel(std::string_view key) {
+  const std::vector<mojom::ModelPtr>& all_models = GetModels();
+
+  auto match_iter = std::find_if(
+      all_models.cbegin(), all_models.cend(),
+      [&key](const mojom::ModelPtr& model) { return model->key == key; });
+  if (match_iter != all_models.cend()) {
+    return &*match_iter->get();
+  }
+
+  return nullptr;
+}
+
+void ModelService::AddCustomModel(mojom::ModelPtr model) {
```

Comment:
Should the key be decided in this function instead of the UI? Advantage is cross-platform.

Reply:
yes, uuid will now be generated here. new commit